controllers/notificationpolicy_controller.go

```diff
@@ -222,6 +344,100 @@ func (r *GrafanaNotificationPolicyReconciler) SetupWithManager(mgr ctrl.Manager)
 			}
 			return requests
 		})).
+		Watches(&grafanav1beta1.GrafanaNotificationPolicyRoute{}, handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, o client.Object) []reconcile.Request {
+			// resync all notification policies that have a routeSelector that matches the routes labels
+			nps := &grafanav1beta1.GrafanaNotificationPolicyList{}
+			if err := r.List(ctx, nps); err != nil {
+				r.Log.Error(err, "failed to fetch notification policies for watch mapping")
+				return nil
+			}
+			requests := []reconcile.Request{}
+			for _, np := range nps.Items {
+				if !hasRouteSelector(np.Spec.Route) {
+					continue
+				}
+
+				if np.GetNamespace() != o.GetNamespace() && !np.IsCrossNamespaceImportAllowed() {
+					continue
+				}
+
+				requests = append(requests,
+					reconcile.Request{
+						NamespacedName: types.NamespacedName{
+							Name:      np.Name,
+							Namespace: np.Namespace,
+						},
+					})
+			}
+			return requests
+		})).
 		WithEventFilter(ignoreStatusUpdates()).
 		Complete(r)
 }
+
+// getMatchingNotificationPolicyRoutes retrieves all GrafanaNotificationPolicyRoutes for the given labelSelector
+// results will be limited to namespace when specified
+func getMatchingNotificationPolicyRoutes(ctx context.Context, k8sClient client.Client, labelSelector *metav1.LabelSelector, namespace *string) (*v1beta1.GrafanaNotificationPolicyRouteList, error) {
+	if labelSelector == nil {
+		return nil, nil
+	}
+
+	var list v1beta1.GrafanaNotificationPolicyRouteList
+	opts := []client.ListOption{
+		client.MatchingLabels(labelSelector.MatchLabels),
+	}
+
+	if namespace != nil {
+		opts = append(opts, client.InNamespace(*namespace))
+	}
+
+	err := k8sClient.List(ctx, &list, opts...)
+	return &list, err
+}
+
+// updateNotificationPolicyRoutesStatus sets status conditions and emits a merged event to all matched notification policy routes
+func (r *GrafanaNotificationPolicyReconciler) updateNotificationPolicyRoutesStatus(ctx context.Context, notificationPolicy *grafanav1beta1.GrafanaNotificationPolicy, routes []*v1beta1.GrafanaNotificationPolicyRoute) error {
+	if notificationPolicy == nil || routes == nil {
+		return nil
+	}
+
+	for _, route := range routes {
+		r.Recorder.Event(route, corev1.EventTypeNormal, "Merged", fmt.Sprintf("Route merged into NotificationPolicy %s/%s", notificationPolicy.GetNamespace(), notificationPolicy.GetName()))
```

Comment:
@theSuess  Are events planned as a replacement instead of status updates or an addition to conditions?
Or is there a reason it has not been used yet?

Reply:
No reason other than it didn't occur to us up until now. I can see this being useful for debugging other resources as well.